### PR TITLE
fix(auth): enable JWT auth on all specified calls

### PR DESCRIPTION
**Pull request by AntoineDao**
pollination/k8s-cluster-management#9

Users will still have to login somewhere else to retrieve a JWT from a UI. I tried using Oauth2 flow with auth0 and openapi 3 and it's UI but I could only retrieve the access_token instead of the id_token (which is the (JWT in this case). I had a quick play with OIDC but couldn't figure out how it was intepreted by OpenApi so gave up. I reckon forcing users to retrieve the JWT from somewhere however will be a good way to introduce them to how they will be using it with Grasshopper.

We can refine this later on of course but for now I reckon this will do. Thoughts?

### CODE REVIEW

File: openapimixer.py [hunk 1]
```diff
@@ -15,7 +15,7 @@ def main(openapi_links):
     
     # add empty fields
     openapi_doc['paths'] = {}
-    openapi_doc['components'] = {'schemas': {}, 'securitySchemes': {}}
+    openapi_doc['components']['schemas'] = {}
```
Comment (mostaphaRoudsari) [inline openapimixer.py]: If we are adding this to `openapi.yaml` as a static value which I would say is a good idea then we don't really need this line.

File: spec/openapi.yaml [hunk 2]
```diff
@@ -37,6 +44,3 @@ tags:
     description: |
       Workflow is a collection of usually interrelated simulations. Simulations will be
       executed in steps or in parallel based on interdependencies between simulations.
-
-security:
```
Comment (mostaphaRoudsari) [inline spec/openapi.yaml]: I might not understand your intent here correctly but this should be kept in order to apply the defined `securitySchemes` , JWT in this case, to all the endpoints. You can also setup the scope which I didn't think we need in this case. See here for more information:  https://swagger.io/docs/specification/authentication/#security

Comment (AntoineDao) [reply spec/openapi.yaml]: The way to think of this is that `securitySchemes` defines the type of authentication mechansims you support and how they are configured. The `security` fields on the other hand are applied at route level to determine which routes require security and what mechanisms can be used and with which scopes. In this case it seems the microservice openapi definition yamls contain the security def within them (i think). Testing locally it all seemed to work as expected so happy to leave it at that for now.

File: openapimixer.py (hunk 1 above)
Comment (AntoineDao) [reply openapimixer.py]: I get an error if I remove this line :stuck_out_tongue:

**Comment by AntoineDao** [comment]
Not exactly. The way to think of it is that we define the authentication mechanism and configuration (eg: which server endpoint to call with OAuth) globally and give those a specific name. We can then apply those schemes to individual routes (which is done at microservice level).

File: spec/openapi.yaml (hunk 2 above)
Comment (mostaphaRoudsari) [reply spec/openapi.yaml]: My understanding is that it is possible both to use `security` keyword to apply the `securityScheme` to the whole API or an individual endpoint/ operation. This is from the link:

Security is described using the `securitySchemes` and security keywords. You use `securitySchemes` to define all security schemes your API supports, then use `security` to apply specific schemes to **the whole API** or **individual operations**. 

> The security fields on the other hand are applied at route level to determine which routes require security and what mechanisms can be used and with which scopes. In this case it seems the microservice openapi definition yamls contain the security def within them (i think).

This is correct but here is my concern for removing it from the combined API documentation. I can imagine that people will use the combined API specification and not the one from each microservice we should keep the security keyword also here so it is applied to the whole API when they generate the client side API.

File: openapimixer.py (hunk 1 above)
Comment (mostaphaRoudsari) [reply openapimixer.py]: Hmm. That is strange. Let's keep it then if that is the case.

File: spec/openapi.yaml (hunk 2 above)
Comment (mostaphaRoudsari) [reply spec/openapi.yaml]: @AntoineDao let's make a decision on this. What are your thoughts? If you are confident that removing these lines is fine then we can keep the PR as is and merge the changes in.

Comment (AntoineDao) [reply spec/openapi.yaml]: I'm 90% sure it's fine. Worse comes to worse we can revert at a later point int time?

**Pull request merged**